packages/foundry/contracts/hooks/OrderHook.sol

```diff
@@ -0,0 +1,476 @@
+        return
+            factory == i_allowedFactory &&
+            IBasePoolFactory(factory).isPoolFromFactory(pool);
+    }
+
+    /*///////////////////////////////////
+    ///////////    HOOK       ///////////
+    ///////////////////////////////////*/
+
+    /// @notice After swap hook
+    /// @dev This hook is called after a swap is executed. We store the previous prices of the tokens in the pool
+    /// @dev This emits an event with the current prices so that we can use that offchain to evalaute the positions of the current active orders
+    /// @dev Only the vault and trusted router is allowed to call this hook
+    /// @param params The parameters of the swap
+    function onAfterSwap(
+        AfterSwapParams calldata params
+    )
+        public
+        override
+        onlyVault
+        returns (bool success, uint256 hookAdjustedAmountCalculatedRaw)
+    {
+        if (params.router != i_trustedRouter) {
+            revert OrderHook__UnauthorizedRouter();
+        }
+
+        emit AfterSwapPrice(
+            params.pool,
+            address(params.tokenIn),
+            address(params.tokenOut),
+            params.tokenInBalanceScaled18,
+            params.tokenOutBalanceScaled18
+        );
+
+        prevPrices[params.pool][params.tokenIn] = params.tokenInBalanceScaled18;
+        prevPrices[params.pool][params.tokenOut] = params
+            .tokenOutBalanceScaled18;
+
+        return (true, params.amountCalculatedRaw);
+    }
+
+    /*///////////////////////////////////
+    ///////    PLACE ORDER       ////////
+    ///////////////////////////////////*/
+
+    /// @notice Place an order in a pool
+    /// @dev The order is placed by transferring the tokenIn to this contract
+    /// @param orderType Type of the order
+    /// @param amountIn Amount of tokenIn
+    /// @param triggerPrice The Price at which the order should be executed
+    /// @param pool Pool address
+    /// @param tokenIn Address of the tokenIn
+    /// @param tokenOut Address of the tokenOut
+    /// @param slippageTolerance Slippage percentage allowed
+    /// @return orderId The id of the order
+    function placeOrder(
+        OrderType orderType,
+        uint256 amountIn,
+        uint256 triggerPrice,
+        address pool,
+        address tokenIn,
+        address tokenOut,
+        uint8 slippageTolerance
+    ) external returns (bytes32 orderId) {
+        if (amountIn == 0) {
+            revert OrderHook__OrderAmountInMustBeMoreThanZero();
+        }
+
+        i_vault.getPoolTokenCountAndIndexOfToken(pool, IERC20(tokenIn));
+        i_vault.getPoolTokenCountAndIndexOfToken(pool, IERC20(tokenOut));
+
+        IERC20(tokenIn).safeTransferFrom(msg.sender, address(this), amountIn);
+
+        orderId = keccak256(
+            abi.encodePacked(msg.sender, orderCount, block.timestamp)
+        );
+
+        orders[orderId] = Order({
+            orderId: orderId,
+            trader: msg.sender,
```

Comment:
Alright, so only the account who places the order can cancel / execute it; is that correct?

Reply:
Yes, that’s correct! Only the account that placed the order has the authority to cancel or execute it.